mconfig

```diff
@@ -846,7 +846,7 @@ if [ "$host" = "unix" ]; then
 	echo "=> generating $RPMSPEC ..."
 	rm -f $sourcedir/$RPMSPEC
 	package_rpm_version="`echo "${short_version}" | tr - .`"
-	sed "s/@PACKAGE_VERSION@/${short_version}/;s/@PACKAGE_RPM_VERSION@/${package_rpm_version}/;s/@PACKAGE_RELEASE@/${release_info}/" \
+	sed "s/@PACKAGE_VERSION@/${package_version}/;s/@PACKAGE_RPM_VERSION@/${package_rpm_version}/;s/@PACKAGE_RELEASE@/${release_info}/" \
```

Comment:
I don't thank that @PACKAGE_VERSION@, which is included in the Source tarball name, should include the release_info.  That is not the normal practice, and it leaves behind a new tarball for every build for those of us who choose a new release for each development build.  The source tarball is given just the version number because releases are only supposed to change packaging and not change the official source.

So I suggest that the Source keyword in dist/rpm/singularity.spec.in go back to using %{version} instead of @PACKAGE_VERSION@, that the Version keyword be assigned @PACKAGE_VERSION@ again, that @PACKAGE_VERSION@ be assigned ${package_rpm_version}, and that @PACKAGE_RPM_VERSION@ be deleted from mconfig.  Then use just the part of $version before a colon from $tarball in scripts/make-dist.sh and from the $(VERSION) used in frags/Makefile.stub.

An additional problem with this PR compared to PR #4558 is that it doesn't do the translation from the standard semantic versioning hyphen to the tilde expected by rpm. That requires a change in both mconfig and mlocal/frags/Makefile.stub.  Test by using mconfig -V 3.5.0-rc.1:1.   

Here is a [patch file](https://github.com/sylabs/singularity/files/3763581/fix4558.patch.txt) with my proposed changes to this PR.

Reply:
In response to a suggestion from @dctrud on slack, I updated the patch to not change hyphen to tilde in the tar file name.  It's not letting me upload another attachment, so here's the new patch inline:
```
diff --git a/mconfig b/mconfig
index 8ee9f18..927e792 100755
--- a/mconfig
+++ b/mconfig
@@ -845,8 +845,8 @@ if [ "$host" = "unix" ]; then
 	RPMSPEC=singularity.spec
 	echo "=> generating $RPMSPEC ..."
 	rm -f $sourcedir/$RPMSPEC
-	package_rpm_version="`echo "${short_version}" | tr - .`"
-	sed "s/@PACKAGE_VERSION@/${package_version}/;s/@PACKAGE_RPM_VERSION@/${package_rpm_version}/;s/@PACKAGE_RELEASE@/${release_info}/" \
+	package_rpm_version="`echo "${short_version}" | tr '-' '~'`"
+	sed "s/@PACKAGE_VERSION@/${short_version}/;s/@PACKAGE_RPM_VERSION@/${package_rpm_version}/;s/@PACKAGE_RELEASE@/${release_info}/" \
 		$sourcedir/dist/rpm/$RPMSPEC.in >$sourcedir/$RPMSPEC
 fi
 
diff --git a/mlocal/frags/Makefile.stub b/mlocal/frags/Makefile.stub
index eba95b8..5982c9d 100644
--- a/mlocal/frags/Makefile.stub
+++ b/mlocal/frags/Makefile.stub
@@ -79,10 +79,10 @@ rpm: dist
 	      --define '_sysconfdir $(RPMPREFIX)/etc' \
 	      --define '_localstatedir $(RPMPREFIX)/var' \
 	      --define '_mandir $(RPMPREFIX)/share/man' \
-	      $(SOURCEDIR)/singularity-$(VERSION).tar.gz; \
+	      $(SOURCEDIR)/singularity-$(SHORT_VERSION).tar.gz; \
 	  else \
 	    rpmbuild $(RPMCLEAN) -ta \
-	      $(SOURCEDIR)/singularity-$(VERSION).tar.gz; \
+	      $(SOURCEDIR)/singularity-$(SHORT_VERSION).tar.gz; \
 	  fi)
 
 .PHONY: cscope
diff --git a/scripts/make-dist.sh b/scripts/make-dist.sh
index 8ae0b21..19d57a7 100755
--- a/scripts/make-dist.sh
+++ b/scripts/make-dist.sh
@@ -25,7 +25,8 @@ fi
 echo " DIST setup VERSION: ${version}"
 echo "${version}" > VERSION
 rmfiles="VERSION"
-tarball="${package_name}-${version}.tar.gz"
+short_version="`echo ${version}|cut -d: -f1`"
+tarball="${package_name}-${short_version}.tar.gz"
 echo " DIST create tarball: $tarball"
 rm -f $tarball
 pathtop="$package_name"
```

Reply:
@DrDaveD, we are conflating issue and implementation, so it's difficult to provide any meaningful feedback as to whether the proposed change is OK or not.

Can you please take a look at [this log](https://drive.google.com/open?id=1IQJLqB7doBsEVlY0NB0GLXOPPNbOyWGL) and let me know how the issue is reflected there?

It's basically running `mconfig -V 1.2.3:4.5.6` and then `make dist` and `make rpm`.

- `make dist` is creating a tarball named `singularity-1.2.3:4.5.6.tar.gz` because that _is_ the version number provided using -V
- `make rpm` results in the creation of an rpm called `singularity-1.2.3-4.5.6.el7.x86_64.rpm` which seems perfectly reasonable to me.

Reply:
The problem in that log is that it is including the release portion, after the colon, in the tarball name, and that's wrong.  It's against common practice, and it leaves a mess of a tarball and build directory behind for every build when the release numbers change frequently.

In addition, your example has no hyphen in the release number so you don't see the difference in translating of hyphens to tilde instead of dot.  Rpm uses tilde to do the same thing as hyphen in semantic versioning (treating it as something coming before the version number), so that's what should be used.

Reply:
Would you be able to write out for the different types of version string that may occur here explictly what:

- tarball name
- rpm Version/Release/filename

... is expected?

I get the hyphen to tilde bit that isn't in @mem's example, but I'm really lost about what combinations of hypen, tilde, colon and their relative positions could be needed.

Reply:
Sure.  With examples:

3.5.0:1.2 -> tarball 3.5.0, rpm Version 3.5.0, Release 1.2
3.5.0-rc.1:2.3 -> tarball 3.5.0, rpm version 3.5.0~rc.1, Release 2.3
3.5.0 -> tarball 3.5.0, rpm Version 3.5.0, Release 1

The tarball name must not change for different releases.  In fact, I'm thinking that singularity should never think of including the release number as part of the version.  I suggest instead of using a colon in -V, have a separate optional -R for the release number and have it only affect the contents of singularity.spec.  Like in the following diff, which is quite a bit simpler than the current code:

```
diff --git a/dist/rpm/singularity.spec.in b/dist/rpm/singularity.spec.in
index de858c6..cee3408 100644
--- a/dist/rpm/singularity.spec.in
+++ b/dist/rpm/singularity.spec.in
@@ -98,7 +98,7 @@ cd $GOPATH/%{singgopath}
 
 # Not all of these parameters currently have an effect, but they might be
 #  used someday.  They are the same parameters as in the configure macro.
-./mconfig -V %{version}-%{release} \
+./mconfig -V %{version} -R %{release} \
         --prefix=%{_prefix} \
         --exec-prefix=%{_exec_prefix} \
         --bindir=%{_bindir} \
diff --git a/mconfig b/mconfig
index 8ee9f18..5e5e3f4 100755
--- a/mconfig
+++ b/mconfig
@@ -54,6 +54,7 @@ if test -z "${package_version}" ; then
 	echo "Failed to get package version. Abort."
 	exit 1
 fi
+package_release=1
 
 with_network=1
 with_suid=1
@@ -88,7 +89,7 @@ usage () {
 
 usage_args () {
 	echo "Usage: ${0##*/} [-vsS] [-P profile] [-b builddir] [-c hstcc] [-C tgtcc]"
-	echo "                [-V version] [-x hstcxx] [-X tgtcxx]"
+	echo "                [-V version] [-R release] [-x hstcxx] [-X tgtcxx]"
 	echo "                [--INSTALLDIR=ARG (see below)]"
 	echo
 	echo "  Makeit general options:"
@@ -103,6 +104,7 @@ usage_args () {
 	echo "     -c   build project with host C \`compiler'"
 	echo "     -C   build project with target C\`compiler'"
 	echo "     -V   build project with given version"
+	echo "     -R   build project with given release (default 1)"
 	echo "     -x   build project with host C++ \`compiler'"
 	echo "     -X   build project with target C++ \`compiler'"
 	echo "     -l   build project with source code language (-l C -l Go ...)"
@@ -367,6 +369,12 @@ while [ $# -ne 0 ]; do
      exit
    fi
    package_version="$2"; shift; shift;;
+  -R)
+   if ! echo "$2" | awk '/^-.*/ || /^$/ { exit 2 }'; then
+     echo "error: option requires an argument: $1"
+     exit
+   fi
+   package_release="$2"; shift; shift;;
   -l)
    if ! echo "$2" | awk '/^-.*/ || /^$/ { exit 2 }'; then
      echo "error: option requires an argument: $1"
@@ -387,15 +395,6 @@ while [ $# -ne 0 ]; do
  esac
 done
 
-# use a colon to provide additional version information such as a
-# release number.
-short_version="`echo "$package_version"|sed 's/:.*//'`"
-release_info="`echo "$package_version"|sed 's/^[^:]*://'`"
-if [ "${short_version}" = "${release_info}" ]; then
-	# $package_version has no colon
-	release_info=1
-fi
-
 # non-option param
 if [ $# != 0 ]; then
 	usage_args
@@ -633,7 +632,6 @@ MANDIR := $mandir
 RPMPREFIX:=
 
 NAME := $package_name
-SHORT_VERSION := $short_version
 VERSION := $package_version
 
 HOSTAR := $hstar
@@ -845,8 +843,8 @@ if [ "$host" = "unix" ]; then
 	RPMSPEC=singularity.spec
 	echo "=> generating $RPMSPEC ..."
 	rm -f $sourcedir/$RPMSPEC
-	package_rpm_version="`echo "${short_version}" | tr - .`"
-	sed "s/@PACKAGE_VERSION@/${package_version}/;s/@PACKAGE_RPM_VERSION@/${package_rpm_version}/;s/@PACKAGE_RELEASE@/${release_info}/" \
+	package_rpm_version="`echo "${package_version}" | tr '-' '~'`"
+	sed "s/@PACKAGE_VERSION@/${package_version}/;s/@PACKAGE_RPM_VERSION@/${package_rpm_version}/;s/@PACKAGE_RELEASE@/${package_release}/" \
 		$sourcedir/dist/rpm/$RPMSPEC.in >$sourcedir/$RPMSPEC
 fi
 
@@ -894,6 +892,9 @@ if test "${host}" = "unix" ; then
 fi
 echo "      ---"
 echo "    - version: $package_version"
+if [ "$package_release" != 1 ]; then
+	echo "    - release: $package_release"
+fi
 
 echo "=> $builddir/Makefile ready, try:"
 echo "   $ cd $builddir"
diff --git a/scripts/make-dist.sh b/scripts/make-dist.sh
index 8ae0b21..a17338a 100755
--- a/scripts/make-dist.sh
+++ b/scripts/make-dist.sh
@@ -33,10 +33,6 @@ ln -sf .. builddir/$pathtop
 rmfiles="$rmfiles builddir/$pathtop"
 trap "rm -f $rmfiles" 0
 
-# XXX(mem): In order to accept filenames with colons in it (because of a
-# version number like x.y.z:1.2.3), pass the --force-local flag to tar.
-# This is understood by GNU tar. If other tar programs (also called
-# "tar") don't, this will need to be fixed.
 (echo VERSION; echo $package_name.spec; git ls-files) | \
     sed "s,^,$pathtop/," |
-    tar --force-local -C builddir -T - -czf "$tarball"
+    tar -C builddir -T - -czf "$tarball"
```

Reply:
I'm afraid that particularly this...

> 3.5.0-rc.1:2.3 -> tarball 3.5.0

... doesn't make sense to me. We use `make dist` to create our release tarballs, and I think we don't want a `3.5.0-rc.1` and `3.5.0` to end up with the same tarball name. They aren't the same code, and shouldn't be able to be confused.

Reply:
tried to edit... but GitHub didn't make this appear....

Otherwise, I quite like the idea of separating the customization of rpm .spec release, from the version. In that case, if we can customize that release however, is there even any need to override the version? It can always be changed by setting an annotated tag on the repo if absolutely necessary... and avoiding overriding it means we shouldn't have ambiguity around what the upstream code was.

Reply:
tried to edit... but GitHub didn't make this appear....

Otherwise, I quite like the idea of separating the customization of rpm .spec release, from the version. In that case, if we can customize that release however, is there even any need to override the version? It can always be changed by setting an annotated tag on the repo if absolutely necessary... and avoiding overriding it means we shouldn't have ambiguity around what the upstream code was.

Reply:
>    3.5.0-rc.1:2.3 -> tarball 3.5.0
>
> ... doesn't make sense to me. 

You're right, I'm sorry.  I should have sad tarball 3.5.0-rc.1

Reply:
> if we can customize that release however, is there even any need to override the version? It can always be changed by setting an annotated tag on the repo if absolutely necessary... and avoiding overriding it means we shouldn't have ambiguity around what the upstream code was.

I'm sorry, I don't understand.  What do you mean by "override the version"?

Reply:
> I'm sorry, I don't understand. What do you mean by "override the version"?

I was thinking it should always be okay to just take the version from the current/closest tag as long as an arbitrary release can be set, but never mind - I take that back. @jmstover has mentioned a situation where there's a need to explicitly set a version with `-V` or `VERSION`, not rely on the closest git tag. I was also thinking of debian package version possibilities in my head, considering that the way our versioning adds `+n-xxxx` form if there are changes after the last tag - but that doesn't work for Fedora / EPEL

Reply:
I think I understand now.

Thinking about the 3.5.0-rc.1 case in particular, what that tells me is that _I do_ need to make a distinction between the singularity version number (3.5.0-rc.1) and the same version expressed in a way that RPM understands it in the same way (3.5.0~rc.1, comes before 3.5.0).

The problem with the naming of the tarball is different. Using Debian terminology, we are dealing with non-pristine tarballs, meaning the tarball used to create singularity-3.5.0-1.1.rpm (by passing -V 3.5.0:1.1 to mconfig) is _different_ from the tarball used to create singularity-3.5.0-1.2.rpm (-V 3.5.0:1.2), and _neither_ corresponds to singularity-3.5.0.tar.gz from our releases on Github.

In the previous paragraph I'm _assuming_ the commands used to create this were:

    mconfig -V 3.5.0-1.2
    make -C builddir dist
    make -C builddir rpm

(as found in the log I provided).

The only possible workflow that makes sense to me is something like this:

    git checkout -b custom-3.5.0 v3.5.0
    # modify stuff
    git commit -m "modified stuff"
    git clean -dxf
    mconfig -V 3.5.0:1.1
    make -C builddir dist
    make -C builddir rpm

This creates an rpm with version "3.5.0" and release info "1.1.something". The tarball is named "3.5.0:1.1" because the modified source code has to be stored _somewhere_.

If you don't pass -V here, the resulting version would have been "3.5.0+1-abcdef", which creates a rpm with version "3.5.0+1-abcdef" and release info "1.something". I do believe this is more truthful in terms of what's being distributed and what the users are installing.

What did I get wrong?

Reply:
And expanding on that last comment regarding _not_ using -V, I would really prefer to nuke it, because as shown here, this is creating a problem, not solving one. As @dctrud suggested, if you wanted to control the version number that singularity reports, what you need to do is:

    git checkout -b custom-3.5.0 v3.5.0
    # modify stuff
    git commit -m "modified stuff"
    git tag --annotate v3.5.0+1.1
    git clean -dxf
    mconfig
    make -C builddir dist
    make -C builddir rpm

and that would cause singularity to report a version number "3.5.0+1.1" and the rpm version would be "3.5.0+1.1" with release info "1.something".

Reply:
rpmbuild assumes pristine tarballs, and assumes that nothing related to releases are included in them.  I believe Debian does too.  With rpm, releases are only for modifying the .spec file.  If somebody has a dirty tarball, it's up to them them to separately modify the Release.  I don't think you should have to worry about people distributing dirty tarballs, people only distribute packages.

rpmbuild leaves a separate directory in the rpmbuild/BUILD directory each time the tarball name changes, because it assumes pristine tarballs that only change when the version changes.  That leaves behind a huge mess during development when there's many new builds and each one gets a separate directory name.  This is why I don't use the default version number picked by mconfig, I always set just a new release number for each build.